File: modules/angular2/src/core/zone/ng_zone.ts
import {ListWrapper, StringMapWrapper} from 'angular2/src/facade/collection';
import {normalizeBlank, isPresent, global, ZoneLike} from 'angular2/src/facade/lang';
import {ObservableWrapper, EventEmitter} from 'angular2/src/facade/async';
import {wtfLeave, wtfCreateScope, WtfScopeFn} from '../profile/profile';

export interface NgZoneZone extends ZoneLike {
  /** @internal */
  _innerZone: boolean;
}

/**
 * Interface for a function with zero arguments.
 */
export interface ZeroArgFunction { (): void; }

/**
 * Function type for an error handler, which takes an error and a stack trace.
 */
export interface ErrorHandlingFn { (error: any, stackTrace: any): void; }

/**
 * Stores error information; delivered via [NgZone.onError] stream.
 */
export class NgZoneError {
  constructor(public error: any, public stackTrace: any) {}
}

/**
 * An injectable service for executing work inside or outside of the Angular zone.
 *
 * The most common use of this service is to optimize performance when starting a work consisting of
 * one or more asynchronous tasks that don't require UI updates or error handling to be handled by
 * Angular. Such tasks can be kicked off via {@link #runOutsideAngular} and if needed, these tasks
 * can reenter the Angular zone via {@link #run}.
 *
 * <!-- TODO: add/fix links to:
 *   - docs explaining zones and the use of zones in Angular and change-detection
 *   - link to runOutsideAngular/run (throughout this file!)
 *   -->
 *
 * ### Example ([live demo](http://plnkr.co/edit/lY9m8HLy7z06vDoUaSN2?p=preview))
 * ```
 * import {Component, View, NgZone} from 'angular2/core';
 * import {NgIf} from 'angular2/common';
 *
 * @Component({
 *   selector: 'ng-zone-demo'.
 *   template: `
 *     <h2>Demo: NgZone</h2>
 *
 *     <p>Progress: {{progress}}%</p>
 *     <p *ngIf="progress >= 100">Done processing {{label}} of Angular zone!</p>
 *
 *     <button (click)="processWithinAngularZone()">Process within Angular zone</button>
 *     <button (click)="processOutsideOfAngularZone()">Process outside of Angular zone</button>
 *   `,
 *   directives: [NgIf]
 * })
 * export class NgZoneDemo {
 *   progress: number = 0;
 *   label: string;
 *
 *   constructor(private _ngZone: NgZone) {}
 *
 *   // Loop inside the Angular zone
 *   // so the UI DOES refresh after each setTimeout cycle
 *   processWithinAngularZone() {
 *     this.label = 'inside';
 *     this.progress = 0;
 *     this._increaseProgress(() => console.log('Inside Done!'));
 *   }
 *
 *   // Loop outside of the Angular zone
 *   // so the UI DOES NOT refresh after each setTimeout cycle
 *   processOutsideOfAngularZone() {
 *     this.label = 'outside';
 *     this.progress = 0;
 *     this._ngZone.runOutsideAngular(() => {
 *       this._increaseProgress(() => {
 *       // reenter the Angular zone and display done
 *       this._ngZone.run(() => {console.log('Outside Done!') });
 *     }}));
 *   }
 *
 *
 *   _increaseProgress(doneCallback: () => void) {
 *     this.progress += 1;
 *     console.log(`Current progress: ${this.progress}%`);
 *
 *     if (this.progress < 100) {
 *       window.setTimeout(() => this._increaseProgress(doneCallback)), 10)
 *     } else {
 *       doneCallback();
 *     }
 *   }
 * }
 * ```
 */
export class NgZone {
  /** @internal */
  _runScope: WtfScopeFn = wtfCreateScope(`NgZone#run()`);
  /** @internal */
  _microtaskScope: WtfScopeFn = wtfCreateScope(`NgZone#microtask()`);

  // Code executed in _mountZone does not trigger the onTurnDone.
  /** @internal */
  _mountZone;
  // _innerZone is the child of _mountZone. Any code executed in this zone will trigger the
  // onTurnDone hook at the end of the current VM turn.
  /** @internal */
  _innerZone;

  /** @internal */
  _onTurnStart: ZeroArgFunction;
  /** @internal */
  _onTurnDone: ZeroArgFunction;
  /** @internal */
  _onEventDone: ZeroArgFunction;
  /** @internal */
  _onErrorHandler: ErrorHandlingFn;

  /** @internal */
  _onTurnStartEvents: EventEmitter<any>;
  /** @internal */
  _onTurnDoneEvents: EventEmitter<any>;
  /** @internal */
  _onEventDoneEvents: EventEmitter<any>;
  /** @internal */
  _onErrorEvents: EventEmitter<any>;

  // Number of microtasks pending from _innerZone (& descendants)
  /** @internal */
  _pendingMicrotasks: number = 0;
  // Whether some code has been executed in the _innerZone (& descendants) in the current turn
  /** @internal */
  _hasExecutedCodeInInnerZone: boolean = false;
  // run() call depth in _mountZone. 0 at the end of a macrotask
  // zone.run(() => {         // top-level call
  //   zone.run(() => {});    // nested call -> in-turn
  // });
  /** @internal */
  _nestedRun: number = 0;

  // TODO(vicb): implement this class properly for node.js environment
  // This disabled flag is only here to please cjs tests
  /** @internal */
  _disabled: boolean;

  /** @internal */
  _inVmTurnDone: boolean = false;

  /** @internal */
  _pendingTimeouts: number[] = [];

  /**
   * @param {bool} enableLongStackTrace whether to enable long stack trace. They should only be
   *               enabled in development mode as they significantly impact perf.
   */
  constructor({enableLongStackTrace}) {
    if (global.zone) {
      this._disabled = false;
      this._mountZone = global.zone;
      this._innerZone = this._createInnerZone(this._mountZone, enableLongStackTrace);
    } else {
      this._disabled = true;
      this._mountZone = null;
    }
    this._onTurnStartEvents = new EventEmitter(false);
    this._onTurnDoneEvents = new EventEmitter(false);
    this._onEventDoneEvents = new EventEmitter(false);
    this._onErrorEvents = new EventEmitter(false);
  }

  /**
   * Sets the zone hook that is called just before a browser task that is handled by Angular
   * executes.
   *
   * The hook is called once per browser task that is handled by Angular.
   *
   * Setting the hook overrides any previously set hook.
   *
   * @deprecated this API will be removed in the future. Use `onTurnStart` instead.
   */
  overrideOnTurnStart(onTurnStartHook: ZeroArgFunction): void {
    this._onTurnStart = normalizeBlank(onTurnStartHook);
  }

  /**
   * Notifies subscribers just before Angular event turn starts.
   *
   * Emits an event once per browser task that is handled by Angular.
   */
  get onTurnStart(): /* Subject */ any { return this._onTurnStartEvents; }

  /** @internal */
  _notifyOnTurnStart(parentRun): void {
    parentRun.call(this._innerZone, () => { this._onTurnStartEvents.emit(null); });
  }

  /**
   * Sets the zone hook that is called immediately after Angular zone is done processing the current
   * task and any microtasks scheduled from that task.
   *
   * This is where we typically do change-detection.
   *
   * The hook is called once per browser task that is handled by Angular.
   *
   * Setting the hook overrides any previously set hook.
   *
   * @deprecated this API will be removed in the future. Use `onTurnDone` instead.
   */
  overrideOnTurnDone(onTurnDoneHook: ZeroArgFunction): void {
    this._onTurnDone = normalizeBlank(onTurnDoneHook);
  }

  /**
   * Notifies subscribers immediately after Angular zone is done processing
   * the current turn and any microtasks scheduled from that turn.
   *
   * Used by Angular as a signal to kick off change-detection.
   */
  get onTurnDone() { return this._onTurnDoneEvents; }

  /** @internal */
  _notifyOnTurnDone(parentRun): void {
    parentRun.call(this._innerZone, () => { this._onTurnDoneEvents.emit(null); });
  }

  /**
   * Sets the zone hook that is called immediately after the `onTurnDone` callback is called and any
   * microstasks scheduled from within that callback are drained.
   *
   * `onEventDoneFn` is executed outside Angular zone, which means that we will no longer attempt to
   * sync the UI with any model changes that occur within this callback.
   *
   * This hook is useful for validating application state (e.g. in a test).
   *
   * Setting the hook overrides any previously set hook.
   *
   * @deprecated this API will be removed in the future. Use `onEventDone` instead.
   */
  overrideOnEventDone(onEventDoneFn: ZeroArgFunction, opt_waitForAsync: boolean = false): void {
    var normalizedOnEventDone = normalizeBlank(onEventDoneFn);
    if (opt_waitForAsync) {
      this._onEventDone = () => {
        if (!this._pendingTimeouts.length) {
          normalizedOnEventDone();
        }
      };
    } else {
      this._onEventDone = normalizedOnEventDone;
    }
  }

  /**
   * Notifies subscribers immediately after the final `onTurnDone` callback
   * before ending VM event.
   *
   * This event is useful for validating application state (e.g. in a test).
   */
  get onEventDone() { return this._onEventDoneEvents; }

  /** @internal */
  _notifyOnEventDone(): void {
    this.runOutsideAngular(() => { this._onEventDoneEvents.emit(null); });
  }

  /**
   * Whether there are any outstanding microtasks.
   */
  get hasPendingMicrotasks(): boolean { return this._pendingMicrotasks > 0; }

  /**
   * Whether there are any outstanding timers.
   */
  get hasPendingTimers(): boolean { return this._pendingTimeouts.length > 0; }

  /**
   * Whether there are any outstanding asynchronous tasks of any kind that are
   * scheduled to run within Angular zone.
   *
   * Useful as a signal of UI stability. For example, when a test reaches a
   * point when [hasPendingAsyncTasks] is `false` it might be a good time to run
   * test expectations.
   */
  get hasPendingAsyncTasks(): boolean { return this.hasPendingMicrotasks || this.hasPendingTimers; }

  /**
   * Sets the zone hook that is called when an error is thrown in the Angular zone.
   *
   * Setting the hook overrides any previously set hook.
   *
   * @deprecated this API will be removed in the future. Use `onError` instead.
   */
  overrideOnErrorHandler(errorHandler: ErrorHandlingFn) {
    this._onErrorHandler = normalizeBlank(errorHandler);
  }

  get onError() { return this._onErrorEvents; }

  /**
   * Executes the `fn` function synchronously within the Angular zone and returns value returned by
   * the function.
   *
   * Running functions via `run` allows you to reenter Angular zone from a task that was executed
   * outside of the Angular zone (typically started via {@link #runOutsideAngular}).
   *
   * Any future tasks or microtasks scheduled from within this function will continue executing from
   * within the Angular zone.
   */
  run(fn: () => any): any {
    if (this._disabled) {
      return fn();
    } else {
      var s = this._runScope();
      try {
        return this._innerZone.run(fn);
      } finally {
        wtfLeave(s);
      }
    }
  }

  /**
   * Executes the `fn` function synchronously in Angular's parent zone and returns value returned by
   * the function.
   *
   * Running functions via `runOutsideAngular` allows you to escape Angular's zone and do work that
   * doesn't trigger Angular change-detection or is subject to Angular's error handling.
   *
   * Any future tasks or microtasks scheduled from within this function will continue executing from
   * outside of the Angular zone.
   *
   * Use {@link #run} to reenter the Angular zone and do work that updates the application model.
   */
  runOutsideAngular(fn: () => any): any {
    if (this._disabled) {
      return fn();
    } else {
      return this._mountZone.run(fn);
    }
  }

  /** @internal */
  _createInnerZone(zone, enableLongStackTrace) {
    var microtaskScope = this._microtaskScope;
    var ngZone = this;
    var errorHandling;

    if (enableLongStackTrace) {
      errorHandling =
          StringMapWrapper.merge(global.Zone.longStackTraceZone,
                                 {onError: function(e) { ngZone._notifyOnError(this, e); }});
    } else {
      errorHandling = {onError: function(e) { ngZone._notifyOnError(this, e); }};
    }

    return zone.fork(errorHandling)
        .fork({
          '$run': function(parentRun) {
            return function() {
              try {
                ngZone._nestedRun++;
                if (!ngZone._hasExecutedCodeInInnerZone) {
                  ngZone._hasExecutedCodeInInnerZone = true;
                  ngZone._notifyOnTurnStart(parentRun);
                  if (ngZone._onTurnStart) {
                    parentRun.call(ngZone._innerZone, ngZone._onTurnStart);
                  }
                }
                return parentRun.apply(this, arguments);
              } finally {
                ngZone._nestedRun--;
                // If there are no more pending microtasks, we are at the end of a VM turn (or in
                // onTurnStart)
                // _nestedRun will be 0 at the end of a macrotasks (it could be > 0 when there are
                // nested calls
                // to run()).
                if (ngZone._pendingMicrotasks == 0 && ngZone._nestedRun == 0 &&
                    !this._inVmTurnDone) {
                  if (ngZone._hasExecutedCodeInInnerZone) {
                    try {
                      this._inVmTurnDone = true;
                      ngZone._notifyOnTurnDone(parentRun);
                      if (ngZone._onTurnDone) {
                        parentRun.call(ngZone._innerZone, ngZone._onTurnDone);
                      }
                    } finally {
                      this._inVmTurnDone = false;
                      ngZone._hasExecutedCodeInInnerZone = false;
                    }
                  }

                  if (ngZone._pendingMicrotasks === 0) {
                    ngZone._notifyOnEventDone();
                    if (isPresent(ngZone._onEventDone)) {
                      ngZone.runOutsideAngular(ngZone._onEventDone);
                    }
                  }
                }
              }
            };
          },
          '$scheduleMicrotask': function(parentScheduleMicrotask) {
            return function(fn) {
              ngZone._pendingMicrotasks++;
              var microtask = function() {
                var s = microtaskScope();
                try {
                  fn();
                } finally {
                  ngZone._pendingMicrotasks--;
                  wtfLeave(s);
                }
              };
              parentScheduleMicrotask.call(this, microtask);
            };
          },
          '$setTimeout': function(parentSetTimeout) {
            return function(fn: Function, delay: number, ...args) {
              var id;
              var cb = function() {
                fn();
                ListWrapper.remove(ngZone._pendingTimeouts, id);
              };
              id = parentSetTimeout.call(this, cb, delay, args);
              ngZone._pendingTimeouts.push(id);
              return id;
            };
          },
          '$clearTimeout': function(parentClearTimeout) {
            return function(id: number) {
              parentClearTimeout.call(this, id);
              ListWrapper.remove(ngZone._pendingTimeouts, id);
            };
          },
          _innerZone: true
        });
  }

  /** @internal */
  _notifyOnError(zone, e): void {
    if (isPresent(this._onErrorHandler) || ObservableWrapper.hasSubscribers(this._onErrorEvents)) {
      var trace = [normalizeBlank(e.stack)];

      while (zone && zone.constructedAtException) {
        trace.push(zone.constructedAtException.get());
        zone = zone.parent;
      }
      if (ObservableWrapper.hasSubscribers(this._onErrorEvents)) {
        ObservableWrapper.callEmit(this._onErrorEvents, new NgZoneError(e, trace));
      }
      if (isPresent(this._onErrorHandler)) {
        this._onErrorHandler(e, trace);
      }
    } else {
      console.log('## _notifyOnError ##');
      console.log(e.stack);
      throw e;
    }
  }
}

File: modules/angular2/src/core/zone/ng_zone.dart
library angular.zone;

import 'dart:async';
import 'package:stack_trace/stack_trace.dart' show Chain;

typedef void ZeroArgFunction();
typedef void ErrorHandlingFn(error, stackTrace);

/**
 * A `Timer` wrapper that lets you specify additional functions to call when it
 * is cancelled.
 */
class WrappedTimer implements Timer {
  Timer _timer;
  ZeroArgFunction _onCancelCb;

  WrappedTimer(Timer timer) {
    _timer = timer;
  }

  void addOnCancelCb(ZeroArgFunction onCancelCb) {
    if (this._onCancelCb != null) {
      throw "On cancel cb already registered";
    }
    this._onCancelCb = onCancelCb;
  }

  void cancel() {
    if (this._onCancelCb != null) {
      this._onCancelCb();
    }
    _timer.cancel();
  }

  bool get isActive => _timer.isActive;
}

/**
 * Stores error information; delivered via [NgZone.onError] stream.
 */
class NgZoneError {
  /// Error object thrown.
  final error;
  /// Either long or short chain of stack traces.
  final List stackTrace;
  NgZoneError(this.error, this.stackTrace);
}

/**
 * A `Zone` wrapper that lets you schedule tasks after its private microtask queue is exhausted but
 * before the next "VM turn", i.e. event loop iteration.
 *
 * This lets you freely schedule microtasks that prepare data, and set an {@link onTurnDone} handler that
 * will consume that data after it's ready but before the browser has a chance to re-render.
 *
 * A VM turn consist of a single macrotask followed 0 to many microtasks.
 *
 * The wrapper maintains an "inner" and "mount" `Zone`. The application code will executes
 * in the "inner" zone unless `runOutsideAngular` is explicitely called.
 *
 * A typical application will create a singleton `NgZone`. The mount zone is the `Zone` where the singleton has been
 * instantiated. The default `onTurnDone` runs the Angular change detection.
 */
class NgZone {
  ZeroArgFunction _onTurnStart;
  ZeroArgFunction _onTurnDone;
  ZeroArgFunction _onEventDone;
  ErrorHandlingFn _onErrorHandler;

  final _onTurnStartCtrl = new StreamController.broadcast(sync: true);
  final _onTurnDoneCtrl = new StreamController.broadcast(sync: true);
  final _onEventDoneCtrl = new StreamController.broadcast(sync: true);
  final _onErrorCtrl =
      new StreamController<NgZoneError>.broadcast(sync: true);

  // Code executed in _mountZone does not trigger the onTurnDone.
  Zone _mountZone;
  // _innerZone is the child of _mountZone. Any code executed in this zone will trigger the
  // onTurnDone hook at the end of the current VM turn.
  Zone _innerZone;

  // Number of microtasks pending from _innerZone (& descendants)
  int _pendingMicrotasks = 0;
  // Whether some code has been executed in the _innerZone (& descendants) in the current turn
  bool _hasExecutedCodeInInnerZone = false;
  // _outerRun() call depth. 0 at the end of a macrotask
  // zone.run(() => {         // top-level call
  //   zone.run(() => {});    // nested call -> in-turn
  // });                      // we should only check for the end of a turn once the top-level run ends
  int _nestedRun = 0;

  bool _inVmTurnDone = false;

  List<Timer> _pendingTimers = [];

  /**
   * Associates with this
   *
   * - a "mount" [Zone], which is a the one that instantiated this.
   * - an "inner" [Zone], which is a child of the mount [Zone].
   *
   * @param {bool} enableLongStackTrace whether to enable long stack trace. They should only be
   *               enabled in development mode as they significantly impact perf.
   */
  NgZone({bool enableLongStackTrace}) {
    _mountZone = Zone.current;

    if (enableLongStackTrace) {
      _innerZone = Chain.capture(() => _createInnerZone(Zone.current),
          onError: _onErrorWithLongStackTrace);
    } else {
      _innerZone = _createInnerZone(Zone.current,
          handleUncaughtError: (Zone self, ZoneDelegate parent, Zone zone,
                  error, StackTrace trace) =>
              _onErrorWithoutLongStackTrace(error, trace));
    }
  }

  /**
   * Sets the zone hook that is called just before Angular event turn starts.
   * It is called once per browser event.
   */
  @Deprecated('Use onTurnStart Stream instead')
  void overrideOnTurnStart(ZeroArgFunction onTurnStartFn) {
    _onTurnStart = onTurnStartFn;
  }

  void _notifyOnTurnStart() {
    this._onTurnStartCtrl.add(null);
  }

  /**
   * Notifies subscribers just before Angular event turn starts.
   *
   * Emits an event once per browser task that is handled by Angular.
   */
  Stream get onTurnStart => _onTurnStartCtrl.stream;

  /**
   * Sets the zone hook that is called immediately after Angular processes
   * all pending microtasks.
   */
  @Deprecated('Use onTurnDone Stream instead')
  void overrideOnTurnDone(ZeroArgFunction onTurnDoneFn) {
    _onTurnDone = onTurnDoneFn;
  }

  /**
   * Notifies subscribers immediately after the Angular zone is done processing
   * the current turn and any microtasks scheduled from that turn.
   *
   * Used by Angular as a signal to kick off change-detection.
   */
  Stream get onTurnDone => _onTurnDoneCtrl.stream;

  void _notifyOnTurnDone() {
    this._onTurnDoneCtrl.add(null);
  }

  /**
   * Sets the zone hook that is called immediately after the last turn in
   * an event completes. At this point Angular will no longer attempt to
   * sync the UI. Any changes to the data model will not be reflected in the
   * DOM. `onEventDoneFn` is executed outside Angular zone.
   *
   * This hook is useful for validating application state (e.g. in a test).
   */
  @Deprecated('Use onEventDone Stream instead')
  void overrideOnEventDone(ZeroArgFunction onEventDoneFn,
      [bool waitForAsync = false]) {
    _onEventDone = onEventDoneFn;

    if (waitForAsync) {
      _onEventDone = () {
        if (_pendingTimers.length == 0) {
          onEventDoneFn();
        }
      };
    }
  }

  /**
   * Notifies subscribers immediately after the final `onTurnDone` callback
   * before ending VM event.
   *
   * This event is useful for validating application state (e.g. in a test).
   */
  Stream get onEventDone => _onEventDoneCtrl.stream;

  void _notifyOnEventDone() {
    this._onEventDoneCtrl.add(null);
  }

  /**
   * Whether there are any outstanding microtasks.
   */
  bool get hasPendingMicrotasks => _pendingMicrotasks > 0;

  /**
   * Whether there are any outstanding timers.
   */
  bool get hasPendingTimers => _pendingTimers.isNotEmpty;

  /**
   * Whether there are any outstanding asychnronous tasks of any kind that are
   * scheduled to run within Angular zone.
   *
   * Useful as a signal of UI stability. For example, when a test reaches a
   * point when [hasPendingAsyncTasks] is `false` it might be a good time to run
   * test expectations.
   */
  bool get hasPendingAsyncTasks => hasPendingMicrotasks || hasPendingTimers;

  /**
   * Sets the zone hook that is called when an error is uncaught in the
   * Angular zone. The first argument is the error. The second argument is
   * the stack trace.
   */
  @Deprecated('Use onError Stream instead')
  void overrideOnErrorHandler(ErrorHandlingFn errorHandlingFn) {
    _onErrorHandler = errorHandlingFn;
  }

  /**
   * Notifies subscribers whenever an error happens within the zone.
   *
   * Useful for logging.
   */
  Stream get onError => _onErrorCtrl.stream;

  /**
   * Runs `fn` in the inner zone and returns whatever it returns.
   *
   * In a typical app where the inner zone is the Angular zone, this allows one to make use of the
   * Angular's auto digest mechanism.
   *
   * ```
   * NgZone zone = <ref to the application zone>;
   *
   * void functionCalledFromJS() {
   *   zone.run(() {
   *     // auto-digest will run after this function is called from JS
   *   });
   * }
   * ```
   */
  dynamic run(fn()) {
    // Using runGuarded() is required when executing sync code with Dart otherwise handleUncaughtError()
    // would not be called on exceptions.
    // see https://code.google.com/p/dart/issues/detail?id=19566 for details.
    return _innerZone.runGuarded(fn);
  }

  /**
   * Runs `fn` in the mount zone and returns whatever it returns.
   *
   * In a typical app where the inner zone is the Angular zone, this allows one to escape Angular's
   * auto-digest mechanism.
   *
   * ```
   * void myFunction(NgZone zone, Element element) {
   *   element.onClick.listen(() {
   *     // auto-digest will run after element click.
   *   });
   *   zone.runOutsideAngular(() {
   *     element.onMouseMove.listen(() {
   *       // auto-digest will NOT run after mouse move
   *     });
   *   });
   * }
   * ```
   */
  dynamic runOutsideAngular(fn()) {
    return _mountZone.run(fn);
  }

  void _maybeStartVmTurn(ZoneDelegate parent) {
    if (!_hasExecutedCodeInInnerZone) {
      _hasExecutedCodeInInnerZone = true;
      parent.run(_innerZone, _notifyOnTurnStart);
      if (_onTurnStart != null) {
        parent.run(_innerZone, _onTurnStart);
      }
    }
  }

  dynamic _run(Zone self, ZoneDelegate parent, Zone zone, fn()) {
    try {
      _nestedRun++;
      _maybeStartVmTurn(parent);
      return parent.run(zone, fn);
    } finally {
      _nestedRun--;
      // If there are no more pending microtasks and we are not in a recursive call, this is the end of a turn
      if (_pendingMicrotasks == 0 && _nestedRun == 0 && !_inVmTurnDone) {
        if (_hasExecutedCodeInInnerZone) {
          // Trigger onTurnDone at the end of a turn if _innerZone has executed some code
          try {
            _inVmTurnDone = true;
            _notifyOnTurnDone();
            if (_onTurnDone != null) {
              parent.run(_innerZone, _onTurnDone);
            }
          } finally {
            _inVmTurnDone = false;
            _hasExecutedCodeInInnerZone = false;
          }
        }

        if (_pendingMicrotasks == 0) {
          _notifyOnEventDone();
          if (_onEventDone != null) {
            runOutsideAngular(_onEventDone);
          }
        }
      }
    }
  }

  dynamic _runUnary(Zone self, ZoneDelegate parent, Zone zone, fn(arg), arg) =>
      _run(self, parent, zone, () => fn(arg));

  dynamic _runBinary(Zone self, ZoneDelegate parent, Zone zone, fn(arg1, arg2),
          arg1, arg2) =>
      _run(self, parent, zone, () => fn(arg1, arg2));

  void _scheduleMicrotask(Zone self, ZoneDelegate parent, Zone zone, fn) {
    _pendingMicrotasks++;
    var microtask = () {
      try {
        fn();
      } finally {
        _pendingMicrotasks--;
      }
    };
    parent.scheduleMicrotask(zone, microtask);
  }

  // Called by Chain.capture() on errors when long stack traces are enabled
  void _onErrorWithLongStackTrace(error, Chain chain) {
    if (_onErrorHandler != null || _onErrorCtrl.hasListener) {
      final traces = chain.terse.traces.map((t) => t.toString()).toList();
      if (_onErrorCtrl.hasListener) {
        _onErrorCtrl.add(new NgZoneError(error, traces));
      }
      if (_onErrorHandler != null) {
        _onErrorHandler(error, traces);
      }
    } else {
      throw error;
    }
  }

  // Outer zone handleUnchaughtError when long stack traces are not used
  void _onErrorWithoutLongStackTrace(error, StackTrace trace) {
    if (_onErrorHandler != null || _onErrorCtrl.hasListener) {
      if (_onErrorHandler != null) {
        _onErrorHandler(error, [trace.toString()]);
      }
      if (_onErrorCtrl.hasListener) {
        _onErrorCtrl.add(new NgZoneError(error, [trace.toString()]));
      }
    } else {
      throw error;
    }
  }

  Timer _createTimer(
      Zone self, ZoneDelegate parent, Zone zone, Duration duration, fn()) {
    WrappedTimer wrappedTimer;
    var cb = () {
      fn();
      _pendingTimers.remove(wrappedTimer);
    };
    Timer timer = parent.createTimer(zone, duration, cb);
    wrappedTimer = new WrappedTimer(timer);
    wrappedTimer.addOnCancelCb(() => _pendingTimers.remove(wrappedTimer));

    _pendingTimers.add(wrappedTimer);
    return wrappedTimer;
  }

  Zone _createInnerZone(Zone zone, {handleUncaughtError}) {
    return zone.fork(
        specification: new ZoneSpecification(
            scheduleMicrotask: _scheduleMicrotask,
            run: _run,
            runUnary: _runUnary,
            runBinary: _runBinary,
            handleUncaughtError: handleUncaughtError,
            createTimer: _createTimer),
        zoneValues: {'_innerZone': true});
  }
}
